spec/dynamoid/transaction_write_spec.rb

```diff
@@ -0,0 +1,414 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+# Dynamoid.config.logger.level = :debug
+
+describe Dynamoid::TransactionWrite do
+
+  describe '.execute' do
+    let(:klass) do
+      new_class(class_name: 'Document') do
+        field :name
+      end
+    end
+
+    let(:klass_with_composite_key) do
+      new_class(class_name: 'Cat') do
+        range :age, :integer
+        field :name
+      end
+    end
+
+    let(:klass_with_callback) do
+      new_class(class_name: 'Dog') do
+        field :name
+        field :says
+        field :says_more
+        before_save { self.says = "woof #{self.says_more}" }
+      end
+    end
+
+    let(:klass_with_validation) do
+      new_class do
+        field :name
+        validates :name, length: { minimum: 4 }
+      end
+    end
+
+    before do
+      klass.create_table
+      klass_with_composite_key.create_table
+      klass_with_callback.create_table
+      klass_with_validation.create_table
+    end
+
+    context 'simple primary key' do
+      it 'creates' do
+        obj1 = klass.new(name: "one")
+        obj2 = klass.new
+        obj3 = nil
+        expect(obj1.persisted?).to eql(false)
+        Dynamoid::TransactionWrite.execute do |txn|
+          txn.create! obj1
+          txn.create! obj2, {name: "two"}
+          obj3 = txn.create! klass, {name: "three"}
+        end
+        expect(obj1.persisted?).to eql(true)
+        obj1_found = klass.find(obj1.id)
+        obj2_found = klass.find(obj2.id)
+        obj3_found = klass.find(obj3.id)
+        expect(obj1_found).to eql(obj1)
+        expect(obj2_found).to eql(obj2)
+        expect(obj3_found).to eql(obj3)
+        expect(obj1_found.name).to eql("one")
+        expect(obj2_found.name).to eql("two")
+        expect(obj3_found.name).to eql("three")
+      end
```

Comment:
Validation tests now check there is no rollback with non-bang saves and that there is a rollback with bang! saves.

Reply:
Yeah, I see. So there are still cases to check that a transaction is rolled back when some of operations fails on DynamoDB side, e.g. deleting/updating not existing item, creation of item which primary key already exists.

The documentation states that:

> DynamoDB rejects the entire TransactWriteItems request if any of the following is true:
> - A condition in one of the condition expressions is not met.
> - An ongoing operation is in the process of updating the same item.
> - There is insufficient provisioned capacity for the transaction to be completed.
> - An item size becomes too large (bigger than 400 KB), a local secondary index (LSI) becomes too large, or a similar validation error occurs because of changes made by the transaction.
> - The aggregate size of the items in the transaction exceeds 4 MB.
> - There is a user error, such as an invalid data format.

So all mentioned by me cases are covered by the first documented scenario - `A condition in one of the condition expressions is not met`. So it makes sense to add a case for each transactional method (create/update/delete etc) to ensure that when a method logically fails on DynamoDB side - the whole transaction is rolled back. So transaction is atomic.